src/miner.erl

```diff
@@ -600,9 +445,229 @@ terminate(Reason, _State) ->
     lager:info("stopping: ~p", [Reason]),
     ok.
 
-%% ==================================================================
-%% Internal functions
-%% =================================================================
+%% ============================================================================
+%% Private
+%% ============================================================================
+
+-spec try_create_block(
+    metadata(),
+    blockchain_txn:txns(),
+    non_neg_integer(),
+    blockchain:blockchain(),
+    swarm_keys()
+) ->
+    create_block_result().
+try_create_block(Metadata, Txns, HBBFTRound, Chain, SwarmKeys) ->
+    %% This can actually be a stale message, in which case we'd produce a block
+    %% with a garbage timestamp. This is not actually that big of a deal, since
+    %% it won't be accepted, but we can short circuit some effort by checking
+    %% for a stale hash.
+    {ok, HashCurr} = blockchain:head_hash(Chain),
+    lager:debug("Metadata ~p, current hash ~p", [Metadata, HashCurr]),
+    N = count_consensus_members(Chain),
+    VotesNeeded = N - ((N - 1) div 3),
+    {_, Hashes} = meta_to_stamp_hashes(Metadata),
+    case hash_check_if_stale(HashCurr, Hashes, VotesNeeded) of
+        {ok, {}} ->
+            {ok, create_block(Metadata, Txns, HBBFTRound, Chain, VotesNeeded, SwarmKeys)};
+        {error, {stale, HashStale, Votes}} ->
+            lager:warning("Stale hash: ~p, picked by ~b nodes.", [HashStale, Votes]),
+            {error, stale_hash};
+        {error, {unexpected_counts, Counts}} ->
+            lager:warning(
+                "got unexpected block hashes in stamp information ~p",
+                [Counts]
+            ),
+            %% XXX May also be [], so not just multiple!
+            {error, multiple_hashes}
+    end.
+
+-spec count_consensus_members(blockchain:blockchain()) -> non_neg_integer().
+count_consensus_members(Chain) ->
+    {ok, ConsensusMembers} =
+        blockchain_ledger_v1:consensus_members(blockchain:ledger(Chain)),
+    length(ConsensusMembers).
+
+-spec hash_check_if_stale(H, [H], C) -> {ok, {}} | {error, E} when
+    E ::  {stale, H, C}
+        | {unexpected_counts, [{H, C}]},
+    H :: blockchain_block:hash(),
+    C :: non_neg_integer().
+hash_check_if_stale(HashCurr, Hashes, VotesNeeded) ->
+    case
+        lists:filter(
+            fun ({_, Votes}) -> Votes >= VotesNeeded end,
+            maps:to_list(miner_util:list_count(Hashes))
+        )
+    of
+        %% We expect every stamp to contain the same block hash:
+        [{HashCurr, _}]      -> {ok, {}};
```

Comment:
can this just be `ok`?  I don't understand the purpose of the empty tuple here.

Reply:
Short answer: yes, it can - no probs, but...

Longer answer: it is a bigger issue than it seems and the solution has to break either intuition or composability and semantic consistency - I chose to break intuition.

Why `{ok, A}` vs `ok`? Compatibility with the general result type `{ok, A} | {error, B}` - so that a function is composable with others that return a Result.

If you accept the virtue of the goal of composability, then you have to choose an `A` that represent nothingness/void. Theoretically it is a unit type and is represented as a 0-tuple in theory, SML, OCaml, Haskell and Rust: https://en.wikipedia.org/wiki/Unit_type

We _could_ also choose some atom, like `ok`, `unit`, `void`, etc., but I find these options even more arbitrary and less satisfying.